### PR TITLE
Simple arrow function parens option

### PR DESCRIPTION
**Pull request by suchipi**
This is based on https://github.com/prettier/prettier/pull/2676 by @rattrayalex. Originally, I was going to rebase that branch and make changes there, but rebasing proved difficult due to unrelated changes in master.

This PR adds an "--arrow-function-parentheses" choice option with two options: "always" and "avoid". Avoid is the same as now (try not to print them if you can). Always is the new option (always print them). Avoid is the default, for backwards-compatibility.

Unlike #2676, this does not include the callbacks option AKA functional. Even though I find it interesting, I don't think there is community/collaborator buy-in or consensus, and I don't want to not add "always" because of that.

This PR also includes a fix that @rattrayalex started in his branch to inline type parameters on test callback functions (`it("very long description", <T>(done) => { /* ... */ })`).

### CODE REVIEW

File: docs/options.md [hunk 1]
```diff
@@ -96,6 +96,19 @@ Put the `>` of a multi-line JSX element at the end of the last line instead of b
 | ------- | ------------------------- | ---------------------------- |
 | `false` | `--jsx-bracket-same-line` | `jsxBracketSameLine: <bool>` |
 
+# Arrow Function Parentheses
```
Comment (wKovacs64) [inline docs/options.md]: This should probably be a `##` heading? 😅

File: src/printer.js [hunk 2]
```diff
@@ -3272,6 +3255,19 @@ function printFunctionParams(path, print, options, expandArg, printTypeParams) {
...
+  if (options.arrowFunctionParentheses === "always") {
+    return false;
+  }
+
+  if (options.arrowFunctionParentheses === "avoid") {
+    const node = path.getValue();
+    return canPrintParamsWithoutParens(node);
+  }
+
+  return false;
```
Comment (rattrayalex) [inline src/printer.js]: Probably worth adding a comment indicating that this should be unreachable

**Comment by rattrayalex** [comment]
Oops, looks like you have some Travis failures (ironically looks like prettier needs to be run; do we have a command for that?).  Also need to fix the docs issue wKovacs64 mentioned.

File: src/printer.js (hunk 2 above)
Comment (lipis) [reply src/printer.js]: or just

```js
function shouldPrintParamsWithoutParens(path, options) {
  if (options.arrowFunctionParentheses === "avoid") {
    const node = path.getValue();
    return canPrintParamsWithoutParens(node);
  }
  // possible comment here to explain it :)
  return false;
}
```

Comment (rattrayalex) [reply src/printer.js]: That works too, though I like the explicitness (and having something to grep for, and having a safe default-fallback).

**Comment by azz** [comment]
Obligatory bikeshed... `arrow-function-parentheses` is a really long name... Should we call it `arrow-parens` like ESLint?

**Comment by azz** [comment]
(Aside: should we continue using milestones, or use release labels?)

**Comment by suchipi** [comment]
Let's use milestones instead of release labels; I forgot that milestones existed, and they're nicer for this kind of thing.

**Comment by suchipi** [comment]
Looks like I'll need to add the option to the website, too.

**Comment by duailibe** [comment]
And we should have a way to coordinate pre-release playground options too...

**Comment by qleguennec** [comment]
Any news on adding a `block` parameter to `arrowParens`, like the `eslint` rule?

**Comment by rattrayalex** [comment]
@qleguennec doubtful; personally, I've been using the "always" option and am much happier with it than I expected. I don't mind the extra parens when the computer does all the typing 😀

Which are you currently using / why is it problematic?

**Comment by lydell** [comment]
@qleguennec You might be able to work around this using [prettier-eslint](https://github.com/prettier/prettier-eslint).

**Pull request merged**